### ISSUE TITLE
Allow "catch (e: Error)" instead of "catch (e: unknown)"

### ISSUE THREAD

**Issue by octogonz** [microsoft/TypeScript#42596]
# Suggestion

TypeScript 3.x forbids type annotations in `catch` blocks:
```ts
try {
  // ...
} catch (e: Error) { // <-- TS1996 "Catch clause variable cannot have a type annotation"
  // ...
}
```

But TypeScript 4.x relaxed this rule to accept `any` or `unknown` only:
```ts
try {
  // ...
} catch (e: Error) { // <-- TS1996 "Catch clause variable type annotation must be 'any' or 'unknown' if specified"
  // ...
}
```
⭐ **Suggestion:**  Could we relax the rule to allow `Error` as well?  

Such a declaration would not accurately describe pathological JavaScript code.  

But here's why it makes sense anyway:
1. In a professional code base, thrown objects always implement the `Error` interface.  We have lint rules that enforce this.  And external packages generally follow this rule as well, at least the kind we'd use for professional work.
2. It's wasteful for every single `catch` block to perform paranoid runtime tests for `instanceof Error`.
3. The TypeScript compiler [doesn't even support](https://github.com/microsoft/TypeScript/issues/42389) `instanceof Error`  for transpiled code.
4. Relaxing this rule won't cause any trouble; if some people really prefer `unknown` they can enable it via a lint rule like [no-implicit-any-catch](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-implicit-any-catch.md) without any involvement from the compiler.

## 🔍 Search Terms

`catch` `instanceof`  `TS1996` `1996`

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by lightrow** [comment]
YES, PLEASE. And ideally also a rule to bring back the TypeScript 3.x behaviour with Error being default type for catch block errors, so we don't have to specify Error type in every try-catch block

**Comment by sanjom** [comment]
Yes! I think that would be a good trade-off for situations where we can be sure to receive an "Error" object!

**Comment by ExE-Boss** [comment]
Duplicate of <https://github.com/microsoft/TypeScript/issues/20024>

**Comment by octogonz** [comment]
@ExE-Boss There is a slight difference between the two issues:  #20024 is asking for arbitrary type annotations for catch blocks,   whereas #42596 is /only/ asking for `catch (e: Error)` only.

The rationale that gave was:
1. In a professional code base, thrown objects always implement the `Error` interface.  We have lint rules that enforce this.  And external packages generally follow this rule as well, at least the kind we'd use for professional work.
2. It's wasteful for every single `catch` block to perform paranoid runtime tests for `instanceof Error`.
3. The TypeScript compiler [doesn't even support](https://github.com/microsoft/TypeScript/issues/42389) `instanceof Error`  for transpiled code.
4. Relaxing this rule won't cause any trouble; if some people really prefer `unknown` they can enable it via a lint rule like [no-implicit-any-catch](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-implicit-any-catch.md) without any involvement from the compiler.

These points apply to `Error` because that interface describes the bare minimum properties that any sane error object should have.  Whereas these points maybe would NOT be convincing for custom error interfaces.

That said it does look like @DanielRosenwasser suggested the same idea in  https://github.com/microsoft/TypeScript/issues/20024#issuecomment-344511199:

> I do like the idea of `catch (err as Error)` because, damn it, if you were going to cast it anyway, maybe we should make your life a little easier while keeping it explicit.

**Comment by MickeyPhoenix** [comment]
I think that `catch (e as Error)` is preferable to `catch (e: Error)`. The latter looks like it's something the type-checker would verify for us (which it's not), while the former looks exactly like the unsafe cast that it is.

The practical result of not having a syntax like this is that people just use the default `any` behavior. And `catch (e)` is not only less visibly a type risk than `catch (e as error)`, it is also dangerously over-flexible as a typing:
- It won't catch typos like `messsage`, and
- It won't catch usage errors like `e.stack.substr(10)`.

Adding `catch (e as Error)` will make code safer, easier to read, and more self-documenting.

**Comment by octogonz** [comment]
It's worth pointing out that the TypeScript code base builds with `useUnknownInCatchVariables=false`. 

Maybe the priority for this problem would be more apparent if the compiler developers had to use their own feature. 😋

**Comment by octogonz** [comment]
> I think that `catch (e as Error)` is preferable to `catch (e: Error)`. The latter looks like it's something the type-checker would verify for us (which it's not), while the former looks exactly like the unsafe cast that it is.

This is a really great suggestion. I'm going to update the issue description to include your idea.

**Comment by MickeyPhoenix** [comment]
Thanks! I wish I could take credit for it, but actually it's from @DanielRosenwasser ([over here](https://github.com/microsoft/TypeScript/issues/20024#issuecomment-344511199)). I just spelled out why I think it's such a great idea. :-)

**Comment by octogonz** [comment]
> It's worth pointing out that the TypeScript code base builds with `useUnknownInCatchVariables=false`.

Setting `useUnknownInCatchVariables=false` is a pretty good workaround for this problem. 👍

Turns out it [requires TypeScript 4.4](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-4.html) however (which was surprising since the problematic validation was introduced in 4.0). As soon as I added the setting to our rig, we started seeing breaks in projects that build using an older compiler. Just something to be aware of.

It would be super awesome if `useUnknownInCatchVariables` could be backported to 4.0.x-4.3.x.

**Comment by eduhsoto** [comment]
I tried this, without setting nothig. I don't know if it's a optimal solution. I use firebase auth.
![code](https://user-images.githubusercontent.com/21205552/216743715-ab71f263-1dca-42ca-b648-900d11326508.png)
